### PR TITLE
Add integration test documentation and negative testing

### PR DESCRIPTION
**Pull request by SabreenKaur**
This change adds documentation describing how to write end to end integration tests using the python Tavern testing framework. This change also adds some negative end to end tests that test the HTTP responses when given incorrect stimuli.

This work contributes to #84

### CODE REVIEW

File: integration-tests/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+- Tests are implemented in yaml files and each new test must have a file name with the folllowing format: `test_<test_name>.tavern.yaml`
+- Tavern tests are run in a separate container (called tavern)
+- Tests will only run sucessfully once the full docker environment of all services and test containers are setup 
+
+## Structure of Tavern test cases
+- Each tavern test has a test name and can consist of several stages
+- Each stage has request information such as a HTTP method, url, headers etc, that you submit a request with
+- Each stage has a response field that is populated with the response you expect for that specific request
+    - For example to validate that the key-value pair `val: none` is present in the response the `json` field needs to be checked:
+    ```
```
Comment (thomas-fossati) [inline integration-tests/README.md:13]:
```suggestion
    ```yaml
```

File: integration-tests/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    response:
+        status_code: 200
+        json:
+            nonce: "{nonce:s}"
+    ```
+- Values from a response can be saved for use in subsequent stages. For example the below saves the Location field in the `header`, into the variable `my-var` which can be used in subsequent stages within the same test:
+    ```
```
Comment (thomas-fossati) [inline integration-tests/README.md:20]:
```suggestion
    ```yaml
```

File: integration-tests/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    response:
+        ....
+        save:
+            headers:
+                my-var: Location
+    ```
+- External functions can be written and used to customise validation: 
+    ```
```
Comment (thomas-fossati) [inline integration-tests/README.md:28]:
```suggestion
    ```yaml
```

File: integration-tests/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        verify_response_with:
+            - function: <file_name_with_function>:<func_name>
+    ```
+
+## Setup integration test environment to run tests locally
+Below are the steps to setup the environment and run the integration tests:
+
+1. Open `integration-tests/docker-compose-integration-tests.yml` file and paste the `command` entry into the tavern service configuration, inline with build and volume configurations:
+
+```
```
Comment (thomas-fossati) [inline integration-tests/README.md:40]:
```suggestion
```yaml
```

File: integration-tests/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+Explaination of volumes: this configuration mounts your local `integration-tests` directory into the container at the location `/integration-tests` in the container. Each time you edit a test file locally, the changes will update in the container. This means you wont have to rebuild the docker containers each time you want edit your test. 
+
+Explaination of command: The purpose of the command `sleep infinity` is to allow the container to stay alive so you can shell into the container and run the tests manually.
+
+NOTE: Remember to remove the `command: sleep infinity` line before commiting your changes!
+
+2. Change to the integration-tests directory locally and setup the docker containers by running the below make command.
+
+```
```
Comment (thomas-fossati) [inline integration-tests/README.md:55]:
```suggestion
```bash
```

File: integration-tests/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+cd services/integration-tests
+make integration-tests-up
+```
+
+3. Use the below command to attach to the logs of the running services to see the live logs.
+
+```
+docker compose --env-file=../deployments/docker/default.env --file=docker-compose-integration-tests.yml logs -f -t
+```
```
Comment (thomas-fossati) [inline integration-tests/README.md:64]: split long lines

```suggestion
```bash
docker compose \
  --env-file=../deployments/docker/default.env \
  --file=docker-compose-integration-tests.yml \
  logs -f -t
```
```

File: integration-tests/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+4. In another shell use this command to shell into the tavern test container (this will give you access to an interactive bash shell inside the tavern docker container).
+
+```
```
Comment (thomas-fossati) [inline integration-tests/README.md:68]:
```suggestion
```bash
```

File: integration-tests/README.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+docker exec -it tavern bash
+```
+
+5. Run the tests using the below command in the shell of the container. Make sure you are in the root directory of the container when running the below command.
+
+```
+PYTHONPATH=$PYTHONPATH:integration-tests py.test integration-tests/ -vv
+```
```
Comment (thomas-fossati) [inline integration-tests/README.md:76]:
```suggestion
```bash
PYTHONPATH=$PYTHONPATH:integration-tests \
  py.test integration-tests/ -vv
```
```

**Review by thomas-fossati — COMMENTED** [review]
LGTM!

a few purely editorial comments on the docs

**Review by thomas-fossati — APPROVED** [review]
🚢 it!

File: .github/workflows/integration-tests.yml [hunk 9]
```diff
@@ -19,16 +19,24 @@ jobs:
...
+      uses: actions/upload-artifact@v3
+      with:
+        name: tavern-logs
+        path: ./tavern-tests.log
+    - name: Check for failures
+      run: grep 'FAILURES' tavern-tests.log | python3 scripts/integ-fail.py
     - name: Save docker compose logs
-      run: docker-compose --env-file=deployments/docker/default.env --file=docker-compose-integration-tests.yml logs -f -t >> integration-tests.log &
+      if: always()
+      run: docker-compose --env-file=deployments/docker/default.env --file=integration-tests/docker-compose-integration-tests.yml logs -f -t --no-color >> container.log &
```
Comment (setrofim) [inline .github/workflows/integration-tests.yml:34]: maybe also `tee` the container.log

File: integration-tests/README.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
+# Integration tests with Tavern
+
+Useful aspects to note before getting started:
+- Tests are implemented in yaml files and each new test must have a file name with the folllowing format: `test_<test_name>.tavern.yaml`
```
Comment (setrofim) [inline integration-tests/README.md:4]:
```suggestion
- Tests are defined in YAML files, and each new test must have a file name with the following format: `test_<test_name>.tavern.yaml`
```

File: integration-tests/README.md [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+- Tavern tests are run in a separate container (called tavern)
```
Comment (setrofim) [inline integration-tests/README.md:5]:
```suggestion
- Tavern tests are run in a separate container (called `tavern`)
```

File: integration-tests/README.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+- Tests will only run sucessfully once the full docker environment of all services and test containers are setup 
+
+## Structure of Tavern test cases
+- Each tavern test has a test name and can consist of several stages
```
Comment (setrofim) [inline integration-tests/README.md:9]:
```suggestion
- Each Tavern test has a test name and can consist of several stages
```

File: integration-tests/README.md [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    ....
+volumes:
+    ....
+command: sleep infinity
+```
+Explaination of volumes: this configuration mounts your local `integration-tests` directory into the container at the location `/integration-tests` in the container. Each time you edit a test file locally, the changes will update in the container. This means you wont have to rebuild the docker containers each time you want edit your test. 
+
+Explaination of command: The purpose of the command `sleep infinity` is to allow the container to stay alive so you can shell into the container and run the tests manually.
+
+NOTE: Remember to remove the `command: sleep infinity` line before commiting your changes!
```
Comment (setrofim) [inline integration-tests/README.md:51]:
```suggestion
:warning: NOTE: Remember to remove the `command: sleep infinity` line before commiting your changes!
```

File: integration-tests/README.md [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    logs -f -t
+```
+
+4. In another shell use this command to shell into the tavern test container (this will give you access to an interactive bash shell inside the tavern docker container).
+
+```bash
+docker exec -it tavern bash
+```
+
+5. Run the tests using the below command in the shell of the container. Make sure you are in the root directory of the container when running the below command.
```
Comment (setrofim) [inline integration-tests/README.md:75]:
```suggestion
5. Run the tests using the following command in the shell of the container. Make sure you are in the root directory of the container when running the command.
```

File: integration-tests/README.md [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+3. Use the below command to attach to the logs of the running services to see the live logs.
+
+```
+docker compose \
+    --env-file=../deployments/docker/default.env \
+    --file=docker-compose-integration-tests.yml \
+    logs -f -t
+```
+
+4. In another shell use this command to shell into the tavern test container (this will give you access to an interactive bash shell inside the tavern docker container).
```
Comment (setrofim) [inline integration-tests/README.md:69]:
```suggestion
4. In another terminal, use this command to shell into the tavern test container (this will give you access to an interactive bash shell inside the tavern docker container).
```

File: integration-tests/README.md [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+NOTE: Remember to remove the `command: sleep infinity` line before commiting your changes!
+
+2. Change to the integration-tests directory locally and setup the docker containers by running the below make command.
+
+```bash
+cd services/integration-tests
+make integration-tests-up
+```
+
+3. Use the below command to attach to the logs of the running services to see the live logs.
```
Comment (setrofim) [inline integration-tests/README.md:60]:
```suggestion
3. Use the following command to attach to the logs of the running services to see the live logs.
```

**Review by setrofim — APPROVED** [review]
